src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -1,17 +1,271 @@
 #include "hs_homeobject.hpp"
+#include "replication_message.hpp"
+#include "replication_state_machine.hpp"
+#include "lib/homeobject_impl.hpp"
+
+SISL_LOGGING_DECL(blobmgr)
 
 namespace homeobject {
 
-BlobManager::Result< blob_id_t > HSHomeObject::_put_blob(ShardInfo const&, Blob&&) {
-    return folly::makeUnexpected(BlobError::UNKNOWN);
+BlobManager::AsyncResult< blob_id_t > HSHomeObject::_put_blob(ShardInfo const& shard, Blob&& blob) {
+    auto& pg_id = shard.placement_group;
+    shared< homestore::ReplDev > repl_dev;
+    {
+        std::shared_lock lock_guard(_pg_lock);
+        auto iter = _pg_map.find(pg_id);
+        if (iter == _pg_map.end()) {
+            LOGERROR("failed to put blob with non-exist pg [{}]", pg_id);
+            return folly::makeUnexpected(BlobError::UNKNOWN_PG);
+        }
+        repl_dev = static_cast< HS_PG* >(iter->second.get())->repl_dev_;
+    }
+
+    RELEASE_ASSERT(repl_dev != nullptr, "Repl dev instance null");
+
+    const uint32_t needed_size = sizeof(ReplicationMessageHeader);
+    auto req = repl_result_ctx< BlobManager::Result< BlobInfo > >::make(needed_size, 512);
+
+    uint8_t* raw_ptr = req->hdr_buf_.bytes;
+    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
+    header->msg_type = ReplicationMessageType::PUT_BLOB_MSG;
+    header->payload_size = 0;
+    header->payload_crc = 0;
+    header->shard_id = shard.id;
+    header->pg_id = pg_id;
+    header->header_crc = header->calculate_crc();
+
+    BlobHeader blob_header{};
+    blob_header.shard_id = shard.id;
+    blob_header.total_size = blob.body.size;
+    blob_header.hash_algorithm = BlobHeader::HashAlgorithm::CRC32;
+    blob_header.compute_blob_hash(blob.body, false /* verify */);
+
+    sisl::sg_list sgs;
+    sgs.size = 0;
+    auto block_size = repl_dev->get_blk_size();
+    auto size = sisl::round_up(sizeof(blob_header), block_size);
+    auto buf_header = iomanager.iobuf_alloc(block_size, size);
+    std::memcpy(buf_header, &blob_header, sizeof(blob_header));
+
+    // Create blob header.
+    sgs.iovs.emplace_back(iovec{.iov_base = buf_header, .iov_len = size});
+    sgs.size += size;
+
+    // Append blob bytes.
+    sgs.iovs.emplace_back(iovec{.iov_base = blob.body.bytes, .iov_len = blob.body.size});
+    sgs.size += blob.body.size;
+
+    // Append metadata and update the offsets and total size.
+    if (!blob.user_key.empty()) {
+        size_t user_key_size = blob.user_key.size();
+        sgs.iovs.emplace_back(iovec{.iov_base = blob.user_key.data(), .iov_len = user_key_size});
+        sgs.size += user_key_size;
+        blob_header.total_size += user_key_size;
```

Comment:
the total size in header is updated here, but it was copied to the buffer at line48 with a stale `total_size`.

Reply:
nice catch

Reply:
removed the total_size. With alignment we need blob size and user meta size in the header.